### ISSUE TITLE
Test failures on alpine linux

### ISSUE THREAD

**Issue by gitfoxi** [ivmai/bdwgc#73]
I built on alpine (musl-libc and bash-based) linux. There's some test failures including a segfault. Any advice on how to debug?

```
================================
   gc 7.4.2: ./test-suite.log
================================

# TOTAL: 15
# PASS:  11
# SKIP:  0
# XFAIL: 0
# FAIL:  4
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: cordtest
==============


FAIL: gctest
============

Switched to incremental mode
Emulating dirty bits with mprotect/signals

FAIL: test_cpp
==============

usage: test_cpp number-of-iterations
Assuming 10 iters
Starting iteration 1

FAIL: disclaim_bench
====================

                        fin. ratio       time/s    time/fin.
```

```
PASS: hugetest
PASS: leaktest
./test-driver: line 107: 22132 Segmentation fault      "$@" > $log_file 2>&1
FAIL: cordtest
./test-driver: line 107: 22123 Segmentation fault      "$@" > $log_file 2>&1
FAIL: gctest
PASS: smashtest
PASS: threadleaktest
PASS: initsecondarythread_test
PASS: threadkey_test
PASS: middletest
PASS: subthreadcreate_test
./test-driver: line 107: 22330 Killed                  "$@" > $log_file 2>&1
FAIL: test_cpp
./test-driver: line 107: 23196 Segmentation fault      "$@" > $log_file 2>&1
FAIL: disclaim_bench
PASS: disclaim_test
PASS: staticrootstest
PASS: realloc_test
```

**Comment by ivmai** [comment]
To simplify debugging: 
* use bdwgc master branch
* add parameters to configure: --enable-gc-assertions --disable-threads
* do not use "incremental mode" (-D GC_DISABLE_INCREMENTAL)
* focus on gctest (test.c) debugging (you could even compile it w/o configure: gcc -I include -D GC_ASSERTIONS -D GC_DISABLE_INCREMENTAL -O0 -g tests/test.c)

**Comment by okuoku** [comment]
It seems gc currently never defines `HAVE_DL_ITERATE_PHDR` on non-glibc Linux platforms. Musl have the one and it is required to work with libc.so properly.

By adding it on `dyn_load.c`, `--enable-gc-assertions --disable-threads` passes `make check`. 

For threading support, [Musl's default stack size is way too small compared to glibc](http://wiki.musl-libc.org/wiki/Functional_differences_from_glibc#Thread_stack_size); at least 80k seems not enough to parallel markers. I think we need 256K-ish here.

I'm not sure why incremental collection is not working on Musl yet..

**Comment by ivmai** [comment]
Please contribute the patch

**Comment by okuoku** [comment]
The hard part is Musl have no `__MUSL__` as the [FAQ](http://wiki.musl-libc.org/wiki/FAQ) says so there is no way to statically add any Musl specific workarounds.

Possible solutions are...

1. Determine `dl_iterate_phdr` existence with `AC_TRY_COMPILE` 
2. Consider `defined(LINUX) && !defined(__GLIBC__) && !defined(__UCLIBC__) && !defined(__dietlibc__)` as Musl
3. [Just let distributors to do the workaround](http://git.alpinelinux.org/cgit/aports/commit/?id=1860ab9f0b474dcc7c9dc18c6f28e542fa51af42) 

I'm now trying to do 2 but just started research around dietlibc..(and pray there is no other libc exists on Linux)

In case try-compile preferred over 2, I will prepare patch against `configure.ac` and `CMakeLists.txt` but I believe we have historically avoided doing any dynamic configuration.

**Comment by ivmai** [comment]
I suggest solution 2.
Related issue: #70

**Comment by ivmai** [comment]
Finally I implemented the 1st approach - see my comment in #148 
I have not solved the issue with small stacks for parallel markers, so --disable-parallel-mark should be passed configure.

If you have some patches not submitted to upstream, you are welcome.

PS. Topics related to musl: #66, #70, #72, #84, #123, #148
